osquery/config/config.cpp

```diff
@@ -61,278 +106,85 @@ Status Config::update(const std::map<std::string, std::string>& config) {
-    }
-    tree.erase("additional_monitoring");
-  }
-
-  for (const auto& item : tree) {
-    // Iterate over each top-level configuration key.
-    auto key = std::string(item.first.data());
-    if (key == "scheduledQueries") {
-      LOG(INFO) << RLOG(903) << "config 'scheduledQueries' is deprecated";
-      for (const auto& node : item.second) {
-        auto query_name = node.second.get<std::string>("name", "");
-        mergeScheduledQuery(query_name, node, conf);
-      }
-    } else if (key == "schedule") {
-      for (const auto& node : item.second) {
-        mergeScheduledQuery(node.first.data(), node, conf);
+    if (tree.count("file_paths") > 0) {
+      auto file_paths = tree.get_child("file_paths");
+      WriteLock wlock(config_files_mutex_);
+      for (const auto& category : file_paths) {
+        for (const auto& path : category.second) {
+          auto pattern = path.first;
+          replaceGlobWildcards(pattern);
+          files_[category.first].push_back(path.first);
+        }
       }
-    } else if (key == "options") {
-      for (const auto& option : item.second) {
-        mergeOption(option, conf);
-      }
-    } else if (key == "file_paths") {
-      for (const auto& category : item.second) {
-        mergeFilePath(key, category, conf);
-      }
-    } else {
-      mergeExtraKey(key, item, conf);
     }
-  }
-
-  return Status(0, "OK");
-}
-
-const pt::ptree& Config::getParsedData(const std::string& key) {
-  if (!Registry::exists("config_parser", key)) {
-    return getInstance().empty_data_;
-  }
-
-  const auto& item = Registry::get("config_parser", key);
-  auto parser = std::static_pointer_cast<ConfigParserPlugin>(item);
-  if (parser == nullptr || parser.get() == nullptr) {
-    return getInstance().empty_data_;
-  }
-
-  return parser->data_;
-}
-
-const ConfigPluginRef Config::getParser(const std::string& key) {
-  if (!Registry::exists("config_parser", key)) {
-    return ConfigPluginRef();
-  }
-
-  const auto& item = Registry::get("config_parser", key);
-  const auto parser = std::static_pointer_cast<ConfigParserPlugin>(item);
-  if (parser == nullptr || parser.get() == nullptr) {
-    return ConfigPluginRef();
-  }
-
-  return parser;
-}
-
-Status Config::getMD5(std::string& hash_string) {
-  // Request an accessor to our own config, outside of an update.
-  ConfigDataInstance config;
 
-  std::stringstream out;
-  try {
-    pt::write_json(out, config.data(), false);
-  } catch (const pt::json_parser::json_parser_error& e) {
-    return Status(1, e.what());
-  }
-
-  hash_string = osquery::hashFromBuffer(
-      HASH_TYPE_MD5, (void*)out.str().c_str(), out.str().length());
-
-  return Status(0, "OK");
-}
-
-void Config::addScheduledQuery(const std::string& name, const tree_node& node) {
-  additionalScheduledQuery(name, node, getInstance().data_);
-}
+    for (const auto& plugin : Registry::all("config_parser")) {
+      auto parser = std::static_pointer_cast<ConfigParserPlugin>(plugin.second);
```

Comment:
If you don't mind, while you're refactoring all the things: change this to a `dynamic_pointer_cast` since we're casting downwards from a `Plugin` to non-virtual-inherited `ConfigParserPlugin`. There's an additional exception to be handled like in this example: https://github.com/facebook/osquery/blob/master/osquery/events/events.cpp#L535.